modules/bgsegm/src/bgfg_gsoc.cpp

```diff
@@ -0,0 +1,1000 @@
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+/**
+ * @file   bgfg_gsoc.cpp
+ * @author Vladislav Samsonov <vvladxx@gmail.com>
+ * @brief  Background Subtraction using Local SVD Binary Pattern. See the following paper:
+ * http://www.cv-foundation.org/openaccess/content_cvpr_2016_workshops/w24/papers/Guo_Background_Subtraction_Using_CVPR_2016_paper.pdf
+ * This file also contains implementation of the different yet better algorithm which is called GSOC, as it was implemented during GSOC and was not originated from any paper.
+ *
+*/
+
+#include "precomp.hpp"
+#include <opencv2/calib3d.hpp>
+#include <iostream>
+
+namespace cv
+{
+namespace bgsegm
+{
+namespace
+{
+
+const float LSBPtau = 0.05f;
+
+#ifdef _MSC_VER
+#include <intrin.h>
+#pragma intrinsic(__popcnt)
+#endif
+
+inline int32_t LSBPDist32(int32_t n) {
+#if defined(__GNUC__) || defined(__clang__)
+    return __builtin_popcount(n);
+#elif defined(_MSC_VER)
+    return __popcnt(n);
+#else
+    // Taken from http://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel
+    n = n - ((n >> 1) & 0x55555555);
+    n = (n & 0x33333333) + ((n >> 2) & 0x33333333);
+    return ((n + ((n >> 4) & 0xF0F0F0F)) * 0x1010101) >> 24;
+    // ---
+#endif
+}
+
+inline float L2sqdist(const Point3f& a) {
+    return a.dot(a);
+}
+
+inline float L1dist(const Point3f& a) {
+    return std::abs(a.x) + std::abs(a.y) + std::abs(a.z);
+}
+
+inline float det3x3(float a11, float a12, float a13, float a22, float a23, float a33) {
+    return a11 * (a22 * a33 - a23 * a23) + a12 * (2 * a13 * a23 - a33 * a12) - a13 * a13 * a22;
+}
+
+inline float localSVD(float a11, float a12, float a13, float a21, float a22, float a23, float a31, float a32, float a33) {
+    float b11 = a11 * a11 + a12 * a12 + a13 * a13;
+    float b12 = a11 * a21 + a12 * a22 + a13 * a23;
+    float b13 = a11 * a31 + a12 * a32 + a13 * a33;
+    float b22 = a21 * a21 + a22 * a22 + a23 * a23;
+    float b23 = a21 * a31 + a22 * a32 + a23 * a33;
+    float b33 = a31 * a31 + a32 * a32 + a33 * a33;
+    const float q = (b11 + b22 + b33) / 3;
+
+    b11 -= q;
+    b22 -= q;
+    b33 -= q;
+
+    float p = std::sqrt((b11 * b11 + b22 * b22 + b33 * b33 + 2 * (b12 * b12 + b13 * b13 + b23 * b23)) / 6);
+
+    if (p == 0)
+        return 0;
+
+    const float pi = 1 / p;
+    const float r = det3x3(pi * b11, pi * b12, pi * b13, pi * b22, pi * b23, pi * b33) / 2;
```

Comment:
consider to use [property](https://en.wikipedia.org/wiki/Determinant#Properties_of_the_determinant) (5) of determinant.